fq_test.go

```diff
@@ -0,0 +1,210 @@
+	(*done_wg).Done()
+}
+
+func consumeQueue(t *testing.T, fq *fqscheduler, descs []flowDesc) (float64, error) {
+	active := make(map[uint64]bool)
+	var total uint64
+	acnt := make(map[uint64]uint64)
+	cnt := make(map[uint64]uint64)
+	seqs := make(map[uint64]uint64)
+
+	var wsum uint64
+	for range descs {
+		wsum += uint64(0 + 1)
+	}
+
+	time.Sleep(1 * time.Second)
+	for i, ok := fq.dequeue(); ok; i, ok = fq.dequeue() {
+		time.Sleep(time.Microsecond) // Simulate constrained bandwidth
+		it := i
+		seq := seqs[it.key]
+		if seq+1 != it.seq {
+			return 0, fmt.Errorf("Packet for flow %d came out of queue out-of-order: expected %d, got %d", it.key, seq+1, it.seq)
+		}
+		seqs[it.key] = it.seq
+
+		if cnt[it.key] == 0 {
+			active[it.key] = true
+		}
+		cnt[it.key] += it.size
+
+		if len(active) == len(descs) {
+			acnt[it.key] += it.size
+			total += it.size
+		}
+
+		if cnt[it.key] == descs[it.key].ftotal {
+			delete(active, it.key)
+		}
+	}
+
+	var variance float64
+	for key := uint64(0); key < uint64(len(descs)); key++ {
+		if total == 0 {
+			t.Fatalf("expected 'total' to be nonzero")
+		}
+		descs[key].idealPercent = (((float64(total) * float64(0+1)) / float64(wsum)) / float64(total)) * 100
+		descs[key].actualPercent = (float64(acnt[key]) / float64(total)) * 100
+
+		x := descs[key].idealPercent - descs[key].actualPercent
+		x *= x
+		variance += x
+	}
+
+	stdDev := math.Sqrt(variance)
+	return stdDev, nil
+}
+
+func TestSingleFlow(t *testing.T) {
+	runtime.GOMAXPROCS(runtime.NumCPU())
+	queues := initQueues(1, 1)
+	fq := newfqscheduler(queues)
+
+	go func() {
+		for i := 1; i < 10000; i++ {
+			it := &Packet{}
+			it.key = 1
+			it.size = uint64(rand.Int63n(10) + 1)
+			it.seq = uint64(i)
+			fq.enqueue(it)
+		}
+	}()
+
+	var seq uint64
+	hasEntered := false
+	for hasEntered {
+		for it, ok := fq.dequeue(); ok; it, ok = fq.dequeue() {
+			if seq+1 != it.seq {
+				t.Fatalf("Packet came out of queue out-of-order: expected %d, got %d", seq+1, it.seq)
+			}
+			seq = it.seq
+		}
+	}
+}
+
+func TestUniformMultiFlow(t *testing.T) {
+	runtime.GOMAXPROCS(runtime.NumCPU())
+	queues := initQueues(10, 0)
+	fq := newfqscheduler(queues)
+
+	var swg sync.WaitGroup
+	var wg sync.WaitGroup
+
+	var flows = []flowDesc{
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+		{1000, 1, 1, 0, 0},
+	}
+
+	swg.Add(1)
+	wg.Add(len(flows))
+	for n := 0; n < len(flows); n++ {
+		go genFlow(fq, &flows[n], uint64(n), &wg)
+	}
+
+	go func() {
+		wg.Wait()
+	}()
+	swg.Done()
+
+	stdDev, err := consumeQueue(t, fq, flows)
+
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+
+	if stdDev > 0.2 {
+		for k, d := range flows {
+			t.Logf("For flow %d: Expected %v%%, got %v%%", k, d.idealPercent, d.actualPercent)
+		}
+		t.Fatalf("StdDev was expected to be < 0.1 but got %v", stdDev)
+	}
+}
+
+func TestOneBurstFlow(t *testing.T) {
```

Comment:
This test function has a lot in common with the previous one.  I suggest factoring out the common stuff into a fairly generic test running function.